CRI: Symlink docker logs to CRI defined log path.

Random-Liu opened the pull request:
This PR symlink docker logs to the CRI defined log path.

I manually test myself, and it works. We still need some more unit test and a node e2e test.

@resouer is working on a node e2e test https://github.com/kubernetes/kubernetes/issues/34661.
For the unit test, the current `MakeFakePodSandbox` and `MakeFakeContainer` is not enough for unit test, because I need to create multiple instances for one container and sandbox to test garbage collection.

I'll send a separate PR to refactor the unit test framework in kuberuntime and finish the unit test.

@yujuhong @feiskyer 
/cc @kubernetes/sig-node

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34858)

<!-- Reviewable:end -->

Random-Liu committed:
Symlink docker logs to CRI defined log path.

Random-Liu commented:
@k8s-bot cri test this.

resouer commented:
node e2e test is https://github.com/kubernetes/kubernetes/pull/34877

Random-Liu commented:
@resouer Thanks!

Random-Liu committed:
Add legacy container log location support.

Random-Liu commented:
Add one more commit to add the legacy container log location support.

feiskyer commented on pkg/kubelet/dockershim/docker_container.go:
```diff
@@ -183,7 +187,20 @@ func (ds *dockerService) CreateContainer(podSandboxID string, config *runtimeApi
+	if err != nil {
+		return fmt.Errorf("failed to start container %q: %v", containerID, err)
+	}
+	// TODO: Should we stop the container if an error occurs during start.
```
I think so. If not, dockershim/kubelet doesn't know there is an error happened.

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -269,7 +277,8 @@ func makeMounts(opts *kubecontainer.RunContainerOptions, container *api.Containe
-		containerLogPath := path.Join(opts.PodContainerDir, cid)
+		containerLogPath := filepath.Join(opts.PodContainerDir, cid)
+		// TODO: We should try to use os interface here.
 		fs, err := os.Create(containerLogPath)
```
Replace this with `m.osInterface.Create`?

Random-Liu replied:
A small problem is that `makeMounts` is not a member function of `m`. We can easily change that, just don't know whether we should make this change just for the unit test.

yujuhong commented:
Reviewed 8 of 14 files at r1.
Review status: 8 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/kubelet/dockershim/docker_container.go, line 240 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858#-KUNw1yvXUsuwL_dnRIZ:-KUNw1yvXUsuwL_dnRI_:b-e4szzs) ([raw file](https://github.com/kubernetes/kubernetes/blob/d125b3bd8f38e22e7173b4d744f52a7bcb5d7143/pkg/kubelet/dockershim/docker_container.go#L240)):*
> ```Go
> 		return "", "", fmt.Errorf("failed to inspect container %q: %v", containerID, err)
> 	}
> 	path, ok := info.Config.Labels[containerLogPathLabelKey]
> ```

If there is no label, maybe we can just return an empty path and don't error out? 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[pkg/kubelet/dockershim/docker_container.go, line 194 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858#-KU3y0mNwZ8R42IT2AOC-r2-194:-KUOoKqe-aj9Sknd655w:b-n8ikxs) ([raw file](https://github.com/kubernetes/kubernetes/blob/2501f1c03cbb8cb56b91370f0b1b78bfde82effd/pkg/kubelet/dockershim/docker_container.go#L194)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> I think so. If not, dockershim/kubelet doesn't know there is an error happened.
</details>
It depends on how important we think the container log file is.
For a container which is running but doesn't have corresponding container log file, do we consider it to be usable or not?

@yujuhong WDYT?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[pkg/kubelet/dockershim/docker_container.go, line 240 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858#-KUNw1yvXUsuwL_dnRIZ:-KUOmkEX9Bqoelr_SzgQ:b-e6sy0k) ([raw file](https://github.com/kubernetes/kubernetes/blob/d125b3bd8f38e22e7173b4d744f52a7bcb5d7143/pkg/kubelet/dockershim/docker_container.go#L240)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> If there is no label, maybe we can just return an empty path and don't error out? 
</details>
It should be fine for `StartContainer`, because we just write the label in `CreateContainer`.
However, for `RemoveContainer`, we should not error out, because we may want to remove old containers and old containers may not have labels.
I'll update the PR to return empty path and deal with the empty path at the caller.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
*[pkg/kubelet/dockershim/docker_container.go, line 194 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858#-KU3y0mNwZ8R42IT2AOC-r2-194:-KUOwB75xXl-nEdGAd_y:b3uzcx4) ([raw file](https://github.com/kubernetes/kubernetes/blob/2501f1c03cbb8cb56b91370f0b1b78bfde82effd/pkg/kubelet/dockershim/docker_container.go#L194)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary>
> It depends on how important we think the container log file is.
> For a container which is running but doesn't have corresponding container log file, do we consider it to be usable or not?
> 
> @yujuhong WDYT?
</details>
Let's do the best-effort cleanup and stop the container here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[pkg/kubelet/dockershim/docker_container.go, line 194 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858#-KU3y0mNwZ8R42IT2AOC-r2-194:-KUP0NcrGp-_9dLSpGwd:bb6rskq) ([raw file](https://github.com/kubernetes/kubernetes/blob/2501f1c03cbb8cb56b91370f0b1b78bfde82effd/pkg/kubelet/dockershim/docker_container.go#L194)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Let's do the best-effort cleanup and stop the container here.
</details>
ACK.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34858)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
@feiskyer @yujuhong Addressed comments. PTAL~

Random-Liu commented:
@k8s-bot cri test this.

feiskyer commented:
LGTM

yujuhong commented:
@Random-Liu let's prioritize this one to unblock the unit test PR :-)

Random-Liu commented:
@yujuhong ACK. Will address this soon.

Random-Liu commented:
@yujuhong Addressed comments.

The change I made:
1) Only gc pod log directories when all sources are ready.
2) Do not stop container when fail to create container log symlink.

Random-Liu commented:
@k8s-bot cri test this.

feiskyer commented:
/lgtm

Random-Liu commented:
Rebase and sent again, apply LGTM based on https://github.com/kubernetes/kubernetes/pull/34858#issuecomment-255536043

cyphar commented:
This is causing us issues with kubernetes-incubator/cri-o#162. Has this problem been resolved, and if not where does the issue lie (the Docker or k8s side)?
